// react-native-autocomplete-dropdown/lib/commonjs/AutocompleteDropdownContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AutocompleteDropdownContextProvider = exports.AutocompleteDropdownContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const AutocompleteDropdownContext = exports.AutocompleteDropdownContext = /*#__PURE__*/_react.default.createContext({
  content: undefined,
  setContent: () => null,
  direction: undefined,
  setDirection: () => null,
  activeInputContainerRef: undefined,
  controllerRef: undefined
});
const AutocompleteDropdownContextProvider = ({
  headerOffset = 0,
  children
}) => {
  const [content, setContent] = (0, _react.useState)();
  const [direction, setDirection] = (0, _react.useState)(undefined);
  const [show, setShow] = (0, _react.useState)(false);
  const [dropdownHeight, setDropdownHeight] = (0, _react.useState)(0);
  const [inputMeasurements, setInputMeasurements] = (0, _react.useState)();
  const [opacity, setOpacity] = (0, _react.useState)(0);
  const [contentStyles, setContentStyles] = (0, _react.useState)(undefined);
  const activeInputContainerRef = (0, _react.useRef)(null);
  const controllerRef = (0, _react.useRef)(null);
  const positionTrackingIntervalRef = (0, _react.useRef)();
  (0, _react.useEffect)(() => {
    if (!inputMeasurements?.height) {
      setOpacity(0);
      return;
    }
    if (dropdownHeight && direction === 'up') {
      setContentStyles({
        top: inputMeasurements.y - dropdownHeight - 10 - headerOffset,
        left: inputMeasurements.x,
        width: inputMeasurements.width
      });
      setOpacity(1);
    } else if (direction === 'down') {
      setContentStyles({
        top: inputMeasurements.y + inputMeasurements.height + 5 + headerOffset,
        left: inputMeasurements.x,
        width: inputMeasurements.width
      });
      setOpacity(1);
    }
  }, [direction, dropdownHeight, headerOffset, inputMeasurements?.height, inputMeasurements?.width, inputMeasurements?.x, inputMeasurements?.y]);
  (0, _react.useEffect)(() => {
    if (content) {
      activeInputContainerRef?.current?.measure((x, y, width, height, pageX, pageY) => {
        setInputMeasurements({
          x: pageX,
          y: pageY,
          width,
          height
        });
        setShow(true);
      });
    } else {
      setInputMeasurements(undefined);
      setDropdownHeight(0);
      setOpacity(0);
      setContentStyles(undefined);
      setShow(false);
    }
  }, [content]);
  (0, _react.useEffect)(() => {
    if (show && !!opacity) {
      positionTrackingIntervalRef.current = setInterval(() => {
        requestAnimationFrame(() => {
          activeInputContainerRef?.current && activeInputContainerRef?.current?.measure((_x, _y, width, height, x, y) => {
            setInputMeasurements(prev => JSON.stringify(prev) === JSON.stringify({
              x,
              y,
              width,
              height
            }) ? prev : {
              x,
              y,
              width,
              height
            });
          });
        });
      }, 16);
    } else {
      clearInterval(positionTrackingIntervalRef.current);
    }
    return () => {
      clearInterval(positionTrackingIntervalRef.current);
    };
  }, [opacity, show]);
  const onLayout = (0, _react.useCallback)(e => {
    setDropdownHeight(e.nativeEvent.layout.height);
  }, []);
  return /*#__PURE__*/_react.default.createElement(AutocompleteDropdownContext.Provider, {
    value: {
      content,
      setContent,
      activeInputContainerRef,
      direction,
      setDirection,
      controllerRef
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.clickOutsideHandlerArea,
    onTouchEnd: () => {
      controllerRef.current?.close();
      controllerRef.current?.blur();
    }
  }, children), !!content && show && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    onLayout: onLayout,
    style: {
      ...styles.wrapper,
      opacity,
      ...contentStyles
    }
  }, content));
};
exports.AutocompleteDropdownContextProvider = AutocompleteDropdownContextProvider;
const styles = _reactNative.StyleSheet.create({
  clickOutsideHandlerArea: {
    flex: 1
  },
  wrapper: {
    position: 'absolute'
  }
});
